### PR TITLE
feat: added download entire db as csv/excel

### PR DESCRIPTION
**Pull request by vasusharma7**
For now :
It downloads entire database as CSV or EXCEL.
Things lacking:
- checking corner cases;
- code cleaning;
- making code more modular;
etc...

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,67 +1,79 @@
...
+  },
+  "dependencies": {
+    "@types/mongodb": "^3.5.27",
+    "blob": "^0.1.0",
+    "excel4node": "^1.7.2",
+    "export-from-json": "^1.3.4",
+    "js2excel": "^1.0.1",
+    "json-2-csv": "^3.7.7",
+    "json-as-xlsx": "^1.1.5",
+    "json2xls": "^0.1.2",
```
Comment (yashshah1) [inline package.json:76]: Do we need all these?

Comment (vasusharma7) [reply package.json]: No, I have deleted the unused ones and committed the changes !

File: package.json [hunk 2]
```diff
@@ -68,12 +68,8 @@
   "dependencies": {
     "@types/mongodb": "^3.5.27",
```
Comment (yashshah1) [inline package.json:69]: Add these to dev dependencies

**Review by yashshah1 — CHANGES_REQUESTED** [review]

File: src/download/download.ts [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+interface document {
+  [key: string]: string | number | boolean | null | document | Array<document>;
+}
+type documents = Array<document>;
+
+interface downloadI {
+  mongoURI: string;
+  database: string;
+  type: string;
```
Comment (yashshah1) [inline src/download/download.ts:17]:
```suggestion
  type: "json" | "csv" | "excel";
```

File: src/download/download.ts [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+}
+
+export default async function downloadDatabase(
+  params: downloadI,
+): Promise<void> {
+  const mongoURI: string = params.mongoURI;
+  const database: string = params.database;
+  const type: string = params.type.toLowerCase();
```
Comment (yashshah1) [inline src/download/download.ts:25]:
```suggestion
  const {mongoURI, database, type} params;
```

File: src/download/download.ts [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    useUnifiedTopology: true,
+  });
+
+  try {
+    console.log("Connecting to Mongo Server......");
+    await client.connect();
+    await client.db("admin").command({ ping: 1 });
+    console.log("Connected successfully to Mongo Server");
+
+    const db: any = await client.db(database);
```
Comment (yashshah1) [inline src/download/download.ts:36]: Avoid any types, `client.db` will assign a type to `db`

File: src/download/download.ts [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    const adminDb = client.db().admin();
+
+    // List all the available databases
+    const fetchDbs = await adminDb.listDatabases();
+    let exists = false;
+    let isEmpty = false;
+    for (const dbObject of fetchDbs.databases) {
+      if (database == dbObject.name) {
```
Comment (yashshah1) [inline src/download/download.ts:44]:
```suggestion
      if (database === dbObject.name) {
```

File: src/download/download.ts [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    console.log("Connected successfully to Mongo Server");
+
+    const db: any = await client.db(database);
+    const adminDb = client.db().admin();
+
+    // List all the available databases
+    const fetchDbs = await adminDb.listDatabases();
+    let exists = false;
+    let isEmpty = false;
+    for (const dbObject of fetchDbs.databases) {
```
Comment (yashshah1) [inline src/download/download.ts:43]: Is there no other way, other than iterating through all of the databases?

File: src/download/download.ts [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+      console.log(`Database ${database} is empty`);
+      return;
+    }
+
+    const collectionsArray: any = await db.listCollections().toArray();
+    if (collectionsArray.length === 0) {
+      console.log(`No Collections found in database ${database}`);
+      return;
+    }
+    const dbDir: string = process.cwd() + "/" + database + "_" + params.type;
```
Comment (yashshah1) [inline src/download/download.ts:64]:
```suggestion
    const dbDir: string = `${database}_${params.type}`;
```

File: src/download/download.ts [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    console.log(`Exporting ${collectionsArray.length} collections as ${type}`);
+    collectionsArray.forEach(
+      async (obj: any): Promise<void> => {
```
Comment (yashshah1) [inline src/download/download.ts:67]: Avoid any.

File: src/download/utils.ts [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+        } catch (err) {
+          console.log(err.message);
+        }
+    });
+    modifiedRecords.push(record);
+  });
+  return modifiedRecords;
+};
+
+export default { preprocessRecords };
```
Comment (yashshah1) [inline src/download/utils.ts:30]: Don't combine named and unnamed exports.
Either `export preprocessRecords` or `export const preprocessRecords = () => {...}`

And import them as `import * as utils from "./utils"`

**Review by yashshah1 — CHANGES_REQUESTED** [review]

File: src/download/download.ts (hunk 5 above)
Comment (vasusharma7) [reply src/download/download.ts:36]: No, actually it doesn't return any error or indication on connecting to a non-existing databases.

File: src/download/download.ts (hunk 7 above)
Comment (vasusharma7) [reply src/download/download.ts:43]: No, actually it doesn't return any error or indication on connecting to a non-existing databases.

**Pull request merged**